pkg/middleware/csrf/csrf.go

```diff
@@ -0,0 +1,138 @@
+	trustedOrigins := cfg.SectionWithEnvOverrides("security").Key("csrf_trusted_origins").Strings(" ")
+
+	for _, header := range additionalHeaders {
+		i.originHeaders[header] = struct{}{}
+	}
+	for _, origin := range trustedOrigins {
+		i.trustedOrigins[origin] = struct{}{}
+	}
+
+	return i
+}
+
+func (i *Implementation) Middleware(logger log.Logger) func(http.Handler) http.Handler {
+	// As per RFC 7231/4.2.2 these methods are idempotent:
+	// (GET is excluded because it may have side effects in some APIs)
+	safeMethods := []string{"HEAD", "OPTIONS", "TRACE"}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// If request has no login cookie - skip CSRF checks
+			if strings.Contains(r.URL.Path, "saml") {
+				fmt.Printf("path: %s\n", r.URL.Path)
+			}
+			if _, err := r.Cookie(i.cfg.LoginCookieName); errors.Is(err, http.ErrNoCookie) {
+				next.ServeHTTP(w, r)
+				return
+			}
+			// Skip CSRF checks for "safe" methods
+			for _, method := range safeMethods {
+				if r.Method == method {
+					next.ServeHTTP(w, r)
+					return
+				}
+			}
+			// Skip CSRF checks for "safe" endpoints
+			for endpoint, _ := range i.safeEndpoints {
+				if r.URL.Path == endpoint {
+					next.ServeHTTP(w, r)
+					return
+				}
+			}
+			// Otherwise - verify that Origin matches the server origin
+			netAddr, err := util.SplitHostPortDefault(r.Host, "", "0") // we ignore the port
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
+			}
+			origins := map[string]struct{}{}
+			for header := range i.originHeaders {
+				origin, err := url.Parse(r.Header.Get(header))
+				if err != nil {
+					logger.Error("error parsing Origin header", "header", header, "err", err)
+				}
+				if origin.String() != "" {
+					origins[origin.Hostname()] = struct{}{}
+				}
+			}
+
+			// No Origin header sent, skip CSRF check.
+			if len(origins) == 0 {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			trustedOrigin := false
+			for o := range i.trustedOrigins {
+				if _, ok := origins[o]; ok {
+					trustedOrigin = true
+					break
+				}
+			}
+
+			_, hostnameMatches := origins[netAddr.Host]
+			if netAddr.Host == "" || !trustedOrigin && !hostnameMatches {
+				http.Error(w, "origin not allowed", http.StatusForbidden)
+				return
+			}
+
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func (i *Implementation) TrustOrigin(origin string) {
+	i.trustedOrigins[origin] = struct{}{}
+}
+
+func (i *Implementation) AddOriginHeader(headerName string) {
+	i.originHeaders[headerName] = struct{}{}
+}
+
+// AddSafeEndpoint is used for endpoints requests to skip CSRF check
+func (i *Implementation) AddSafeEndpoint(endpoint string) {
+	// TODO: force trimprefix or add prefix.
+	// unclear when we have one or the other
+	endpoint = strings.TrimPrefix(endpoint, "/")
+	i.safeEndpoints[endpoint] = struct{}{}
+}
```

Comment:
@alexanderzobnin can you comment on this. Sometimes we have a prefix slash and sometimes not in the `r.URL.Path`

Reply:
Maybe we also can trim trailing slash? I'm a bit concern about removing prefix since `/foo/bar` and `foo/bar` are technically different paths. Could you give some examples when path does not contain prefix slash?

Reply:
after discussion with alex, if we have any functionality of editing the safeEndpoints, that might be a security issue. We remove this functionality